### PR TITLE
some fixes in type generation

### PR DESCRIPTION
**Pull request by vankop**
- fix bug with getting type from symbol
- spread argument can not be optional
- fix symbol type detection

this should allow to upgrade ts version to 4.5.5

### CODE REVIEW

File: generate-types/index.js [hunk 1]
```diff
@@ -731,7 +732,7 @@ const printError = (diagnostic) => {
 				}
 				const flags = prop.getFlags();
 				if (name.startsWith("__@")) {
-					name = `[Symbol.${name.slice(3)}]`;
+					name = `[Symbol.${name.slice(3, name.lastIndexOf('@'))}]`;
```
Comment (vankop) [inline generate-types/index.js:735]: symbols now are named like `__@iterator@65`

File: generate-types/index.js [hunk 2]
```diff
@@ -1847,9 +1847,10 @@ const printError = (diagnostic) => {
 							code: getCode(arg.type, innerTypeArgs),
 							optional: false,
 					  };
-				canBeOptional = canBeOptional && optional;
+				const isCurrentOptional = optional && !arg.spread;
```
Comment (vankop) [inline generate-types/index.js:1850]: there is a case in types where spread argument is optional, not sure why right now all works

File: generate-types/index.js [hunk 3]
```diff
@@ -877,7 +878,6 @@ const printError = (diagnostic) => {
 		}
 
 		if (type.intrinsicName) {
-			if (type.intrinsicName === "error") debugger;
```
Comment (vankop) [inline generate-types/index.js:880]: same pain...

File: generate-types/index.js [hunk 4]
```diff
@@ -717,7 +718,7 @@ const printError = (diagnostic) => {
 				if (!isIncluded(nameForFilter)) {
 					if (name.startsWith("_") && !name.startsWith("__")) continue;
 				}
-				if (name.startsWith("__@") && !/^__@[^@]+$/.test(name)) continue;
+				if (name.startsWith("__@") && (prop.flags & ts.TypeFlags.ESSymbolLike) === 0) continue;
```
Comment (vankop) [inline generate-types/index.js:721]: this skips `Symbol.toStringTag` for unknown reason.. looks like TS does not know that `Symbol.toStringTag` is "es" symbol.. 

I did not dive in, looks like we are ok with not exposing `[Symbol.toStringTag](): string`

File: generate-types/index.js [hunk 5]
```diff
@@ -351,6 +351,7 @@ const printError = (diagnostic) => {
 					return type;
 				}
 			}
+			if (symbol.type) return symbol.type;
```
Comment (vankop) [inline generate-types/index.js:354]: `symbol.type` not always exists

**Comment by vankop** [comment]
Also if run this on webpack types there is a case here
```js
/**
 * @callback ReportProgress
 * @param {number} p
 * @param {...string[]} [args]
 * @returns {void}
 */

/** @type {WeakMap<Compiler,ReportProgress>} */
const progressReporters = new WeakMap();

class ProgressPlugin {
	/**
	 * @param {Compiler} compiler the current compiler
	 * @returns {ReportProgress} a progress reporter, if any
	 */
	static getReporter(compiler) {
		return progressReporters.get(compiler);
	}
}
```
`getReporter` start returning callback with `args: ...string[][]` this is fixed with
```js
/**
 * @callback ReportProgress
 * @param {number} p
 * @param {...string} [args]
 * @returns {void}
 */
```

**Review by alexander-akait — APPROVED** [review]

**Pull request merged**